### PR TITLE
diophantine: make trivial solution satisfy assumptions

### LINKED ISSUE #26783: diophantine and diop_solve ignore assumptions
Using `solvers.diophantine` `diophantine` and `diop_solve` functions with integer and positive assumed symbols, also `{(0, 0)}` and negative solutions are returned.
For example:
```python
from sympy import Symbol
from sympy.solvers.diophantine import diophantine, diop_solve

x = Symbol("x", integer=True, positive=True)
y = Symbol("y", integer=True, positive=True)
eq = 10*x**2 + 5*x*y - 3*y
print(diophantine(eq))
print(diop_solve(eq))
```
prints
```
{(0, 0)}
{(1, -5), (-3, 5), (0, 0)}
```
while it should print at least an empty solution set.

### PR DESCRIPTION
**Pull request by zylipku**
Currently, trivial solutions will always be returned regardless of the assumptions of positiveness. An additional check is now added after a trivial solution is appended.

#### References to other Issues or PRs

Fixes #26783 

#### Brief description of what is fixed or changed
* An assumption check for the trivial solution is added.
* 3 more test cases in the test function.
* Changes have been made to `diop_solve()` to warn users that the function will not filter out infeasible solutions.

#### Other comments
I'm not sure if the additional warning is enough for users who care about the variable assumptions. Maybe we should deprecate the `diop_solve()`, which forces the users to directly use `diophantine()` instead?

#### Release Notes

* solvers
  * Now `diophantine()` only returns a trivial solution under assumptions.

### CODE REVIEW

File: sympy/solvers/diophantine/diophantine.py [hunk 1]
```diff
@@ -1482,7 +1482,14 @@ def diophantine(eq, param=symbols("t", integer=True), syms=None,
     null = tuple([0]*len(var))
     # if there is no solution, return trivial solution
     if not sols and eq.subs(zip(var, null)).is_zero:
-        sols.add(null)
+        null_feasible = True
+        for val, symb in zip(null, var):
+            if check_assumptions(val, **symb.assumptions0) is False:
+                null_feasible = False
+                break
+        if null_feasible:
```
Comment (smichr) [inline sympy/solvers/diophantine/diophantine.py:1490]:
```suggestion
        if all(check_assumptions(v, **s.assumptions0) is not False for v, s in zip(mull, var)):
```

**Comment by smichr** [comment]
> Maybe we should deprecate the `diop_solve()`

`diop_solve` need not be deprecated -- the helper functions for `diophantine` just do their job of returning solutions and `diophantine` can check or do permutations as desired.

**Comment by zylipku** [comment]
I understand that `diop_solve()` is necessary for solving the equations. Actually I'm thinking if should add a warning message or something for these helper functions since new users are easy to confuse them with the basic `diophantine()` and thus expect the same behaviors.

**Comment by smichr** [comment]
> should add a warning message or something for these helper functions since new users are easy to confuse them with the basic `diophantine()`

If the users check the documentation, it notes at the outset, "This module contains `diophantine` and helper functions that are needed to solve certain Diophantine equations." If you want to add a statement indicating that the helper functions are not intended for casual use, that would be ok. 

I didn't think that `diop_solve` was imported into the public space, but it is: `__init__` has "'diophantine', 'classify_diop', 'diop_solve'" as being imported by "*".

**Comment by smichr** [comment]
I think the note you added is already helpful. Do you think more is necessary?

Waiting to commit until I hear your thoughts.

**Comment by zylipku** [comment]
> I didn't think that `diop_solve` was imported into the public space, but it is: `__init__` has "'diophantine', 'classify_diop', 'diop_solve'" as being imported by "*".

Do you think we need to remove `diop_solve()` from the `__init__` file? This may discourage users from calling it directly unless they know what they are doing. 

Now all the functionality of `diop_solve()` can be realized by `diophantine()` (please correct me if I miss any details), so I believe the function currently satisfies the deprecation policy:
> With that being said, it may be decided that the API of SymPy must change in some incompatible way. Some reasons APIs are changed can include: 
> * The existing API is confusing.
> * ... ...

By the way, in the docs, `diop_solve()` is not classified as "User Functions". 
https://github.com/sympy/sympy/blob/a4a599d169656e2072b316aa4b4646b3dad33422/sympy/solvers/diophantine/diophantine.py#L36-L37
I think either the `__init__` file you've mentioned or the diophantine documentation needs to be changed to avoid any confusion.

**Comment by smichr** [comment]
Caution about using `diop_solve` is provided in `diophantine`. And your note in `solve_diop` reminds the user that the assumptions on symbols are not checked. I also learned that `diop_solve` is only loaded into the namespace when doing `from sympy.solvers.diophantine import *` not when doing `from sympy import *`. So I think this is ok.

File: sympy/solvers/diophantine/diophantine.py [hunk 2]
```diff
@@ -1558,7 +1560,8 @@ def diop_solve(eq, param=symbols("t", integer=True)):
 
     Use of ``diophantine()`` is recommended over other helper functions.
     ``diop_solve()`` can return either a set or a tuple depending on the
-    nature of the equation.
+    nature of the equation. Solutions which DO NOT satisfy the assumptions
+    such as `positive=True` are returned as well.
```
Comment (smichr) [inline sympy/solvers/diophantine/diophantine.py:1564]:
```suggestion
    nature of the equation. All non-trivial solutions are returned: assumptions
    on symbols are ignored.
```

**Comment by zylipku** [comment]
Yes, I agree with you. I think the commit is now ready to pull.

**Pull request merged**